janus_server/src/aggregator/accumulator.rs

```diff
@@ -0,0 +1,165 @@
+    }
+
+    /// Update the in-memory accumulators with the provided output share and report nonce.
+    pub(super) fn update(
+        &mut self,
+        output_share: &A::OutputShare,
+        nonce: Nonce,
+    ) -> Result<(), datastore::Error> {
+        let key = nonce
+            .time()
+            .to_batch_unit_interval_start(self.min_batch_duration)
+            .map_err(|e| datastore::Error::User(e.into()))?;
+        if let Some(accumulate) = self.accumulations.get_mut(&key) {
+            accumulate
+                .update(output_share, nonce)
+                .map_err(|e| datastore::Error::User(e.into()))?;
+        } else {
+            self.accumulations.insert(
+                key,
+                Accumulate {
+                    aggregate_share: A::AggregateShare::from(output_share.clone()),
+                    report_count: 1,
+                    checksum: NonceChecksum::from_nonce(nonce),
+                },
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Write the accumulated aggregate shares, report counts and checksums to the datastore. If a
+    /// batch unit aggregation already exists for some accumulator, it is updated. If no batch unit
+    /// aggregation exists, one is created and initialized with the accumulated values.
+    #[tracing::instrument(skip(self, tx), err)]
+    pub(super) async fn flush_to_datastore<E>(
+        self,
+        tx: &Transaction<'_>,
+    ) -> Result<(), datastore::Error>
+    where
+        for<'a> A::AggregateShare: TryFrom<&'a [u8], Error = E>,
+        E: std::fmt::Display,
+    {
+        for (unit_interval_start, accumulate) in self.accumulations {
+            let unit_interval = Interval::new(unit_interval_start, self.min_batch_duration)?;
+
+            let mut batch_unit_aggregations = tx
+                .get_batch_unit_aggregations_for_task_in_interval::<A, _>(
+                    self.task_id,
+                    unit_interval,
+                    &self.aggregation_param,
+                )
+                .await?;
+
+            if batch_unit_aggregations.len() > 1 {
+                return Err(datastore::Error::DbState(format!(
+                    "found {} batch unit aggregation rows for task {}, interval {unit_interval}, agg parameter {:?}",
+                    batch_unit_aggregations.len(),
+                    self.task_id,
+                    self.aggregation_param,
+                )));
+            }
+
+            if let Some(batch_unit_aggregation) = batch_unit_aggregations.first_mut() {
+                debug!(
+                    unit_interval_start = ?unit_interval.start(),
+                    "accumulating into existing batch_unit_aggregation_row",
+                );
+                batch_unit_aggregation
+                    .aggregate_share
+                    .merge(&accumulate.aggregate_share)
+                    .map_err(|e| datastore::Error::User(e.into()))?;
+                batch_unit_aggregation.report_count += 1;
```

Comment:
This update is incorrect
```suggestion
                batch_unit_aggregation.report_count += accumulate.report_count;
```